### PR TITLE
Properly check for key equality while using a cached Realm instance

### LINKED ISSUE #901: Wrong Key Exception while multiple access to encrypted Realm file
Hi guys!
I happened to stumble across bizarre situation while accessing encrypted database multiple times using different instances of Realm - the 'Wrong key used to decrypt Realm.' exception was thrown. It was weird because I was sure that i was passing matching byte array. (Of course there was no problem opening not encrypted database multiple times) . It occurs that exception is thrown when different instances of byte arrays are passed to Realm.getInstance(...), due to checking equality by checking references:
 
```
if (realm.key != key) {
   throw new IllegalStateException(DIFFERENT_KEY_MESSAGE);
}
```

My question is: is this behaviour planned? It does not seem much issue, but for me, it seems quite misleading to pass matching byte array, while getting 'Wrong key' exception or modify key byte array on the same reference, but still be able to access Realm.

The code to reproduce the situation:
```
@Override
    protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
        Realm.deleteRealmFile(this, "test");
        byte[] key = new byte[64];
        byte[] key2 = new byte[64];
        Realm realm = Realm.getInstance(this, "test", key);
        key[0] = 0xF; <-the most misleading fragment
        realm = Realm.getInstance(this, "test", key); <-opens
        realm = Realm.getInstance(this, "test", key2); <-Exception
    }
```

### PR DESCRIPTION
**Pull request by emanuelez**
Fixes #901

### CODE REVIEW

**Comment by emanuelez** [comment]
Changelog: 
 * Fixed problem occurring when opening an encrypted Realm with two different instances of the same key.

**Comment by emanuelez** [comment]
@cmelchior @kneth @bmunkholm

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 1]
```diff
@@ -1316,6 +1316,39 @@ private void fileCopy(File src, File dst) throws IOException {
...
+                if (realm2 != null) {
+                    realm2.close();
+                }
+            }
+        } finally {
+            if (realm1 != null) {
+                realm1.close();
+            }
+        }
+
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Remove empty line

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 2]
```diff
@@ -1316,6 +1316,39 @@ private void fileCopy(File src, File dst) throws IOException {
         outStream.close();
     }
 
+    public void testOpeningOfEncryptedRealmWithDifferentKeyInstances() throws Exception {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: There is no checked exceptions being thrown?

**Comment by cmelchior** [comment]
Apart from comments :+1:

**Pull request merged**